src/main/assets/js/session-timeout.ts

```diff
@@ -3,56 +3,87 @@ import { throttle } from 'lodash';
-};
-
-const pingUserActive = throttle(
-  () => {
-    fetch('/active').then(() => {
-      setSaveTimeout();
-    });
-  },
-  eventTimer,
-  { trailing: false }
-);
-
-const getSessionTimeoutInterval = (): number => {
-  const twelveHours: number = 12 * 60 * 60 * 1000;
-  const twentyMinutes: number = 20 * 60 * 1000;
-
-  const timeoutParam = new URL(location.href).searchParams.get('timeout');
-  if (process.env.NODE_ENV === 'development' && timeoutParam && !isNaN(parseInt(timeoutParam))) {
-    return parseInt(timeoutParam) > twentyMinutes ? twentyMinutes : Math.abs(parseInt(timeoutParam));
+
+  scheduleSignOut(): void {
+    this.scheduleNotificationPopup();
+    clearTimeout(this.timeout);
+    this.timeout = setTimeout(this.saveSessionAndRedirect, this.sessionTimeoutInterval);
   }
 
-  return [WEBCHAT_URL, TIMED_OUT_URL].includes(window.location.pathname as PageLink) ? twelveHours : twentyMinutes;
-};
+  scheduleNotificationPopup(): void {
+    clearTimeout(this.notificationTimer);
+    this.notificationTimer = setTimeout(
+      () => this.showNotificationPopup(true),
+      this.sessionTimeoutInterval - this.TIMEOUT_NOTICE
+    );
+  }
+
+  showNotificationPopup(visible: boolean): void {
+    if (visible) {
+      this.notificationPopup?.removeAttribute('hidden');
+    } else {
+      this.notificationPopup?.setAttribute('hidden', 'hidden');
+    }
+  }
+
+  async saveSessionAndRedirect(): Promise<void> {
+    if (this.form) {
+      const formData = new FormData(this.form as HTMLFormElement);
+      formData.append('saveBeforeSessionTimeout', 'true');
+      const url = window.location.pathname;
+      const csrf = formData.get('_csrf') as string;
+      await fetch(url, {
+        method: 'POST',
+        headers: { 'csrf-token': csrf },
+        body: new URLSearchParams(formData as unknown as Record<string, string>),
+      });
+    }
+    window.location.href = `${TIMED_OUT_URL}?lng=${document.documentElement.lang}`;
+  }
+
+  pingUserActive(): void {
+    throttle(() => fetch('/active').then(this.scheduleSignOut), eventTimer, { trailing: false });
+  }
+
+  getSessionTimeoutInterval(): number {
+    const timeoutParam = new URL(location.href).searchParams.get('timeout');
+    if (process.env.NODE_ENV === 'development' && timeoutParam && !isNaN(parseInt(timeoutParam))) {
+      return parseInt(timeoutParam) > this.TWENTY_MINUTES ? this.TWENTY_MINUTES : Math.abs(parseInt(timeoutParam));
+    }
+
+    return [WEBCHAT_URL, TIMED_OUT_URL].includes(window.location.pathname as PageLink)
+      ? this.TWELVE_HOURS
+      : this.TWENTY_MINUTES;
+  }
+}
+
+const sessionTimeout = new SessionTimeout();
 
 setTimeout(() => {
-  ['click', 'touchstart', 'mousemove', 'keypress'].forEach(evt => document.addEventListener(evt, pingUserActive));
+  ['click', 'touchstart', 'mousemove', 'keypress'].forEach(evt =>
+    document.addEventListener(evt, sessionTimeout.pingUserActive)
+  );
 }, eventTimer);
-setSaveTimeout();
+
+sessionTimeout.schedule();
```

Comment:
Looks good from what I can see @benouaer. Something I'm thinking about and can't tell immediately from the code is when the notification pop up comes up, and then the user moves their mouse, will the timer be reset to 20 mins even though the window says you will be timed out in 2min?

Reply:
I will tentatively approve but I think the priority should be a lot of testing with a focus on edge cases.

Reply:
I've disabled the activity pings when the popup is visible so that the user would have to actively extend his session by clicking the button as instructed, otherwise they would be signed out